File: frontend/src/api/reportSearch.ts
import { apiGet, apiPost, apiPut } from './request'

export interface SearchDepartment {
  code: string
  name: string
}

export interface SearchDataSource {
  key: string
  name: string
  enabled: boolean
}

export interface SearchOptions {
  current_department: SearchDepartment
  data_sources: SearchDataSource[]
  default_start_time: string
  default_end_time: string
}

export interface SearchClassificationItem {
  code: string
  name: string
}

export interface SearchClassificationResponse {
  source: string
  level: 'category' | 'type' | 'detail'
  items: SearchClassificationItem[]
}

export interface StatisticsDictionarySource {
  source: 'jjd_jjd' | 'fkd_fkd'
  name: string
  categories: SearchClassificationItem[]
  types: SearchClassificationItem[]
  details: SearchClassificationItem[]
  disabled: {
    category: string[]
    type: string[]
    detail: string[]
  }
}

export interface StatisticsDictionaryConfig {
  sources: StatisticsDictionarySource[]
}

export interface SearchMetric {
  key: string
  label: string
  description: string
  default: boolean
}

export interface SearchMetrics {
  source: string
  dimensions: SearchMetric[]
  measures: SearchMetric[]
}

export type SearchAnalysisType = 'standard' | 'jurisdiction_yoy_summary'

export interface SearchQuery {
  source: 'jjd_jjd' | 'fkd_fkd'
  analysis_type?: SearchAnalysisType
  start_time: string
  end_time: string
  category_codes: string[]
  type_codes: string[]
  detail_codes: string[]
  dimensions: string[]
  measures: string[]
  limit: number
}

export interface SearchResultColumn {
  key: string
  label: string
  type: 'text' | 'number' | 'datetime'
}

export type ReportQueryBlockMode = 'snapshot' | 'dynamic'

export interface ReportQueryBlock {
  id: string
  mode: ReportQueryBlockMode
  query: Omit<SearchQuery, 'start_time' | 'end_time'>
  title: string
  result?: SearchResult | null
  last_updated_at?: string | null
  error?: string | null
}

export interface SearchResult {
  source: SearchDataSource
  department: SearchDepartment
  columns: SearchResultColumn[]
  rows: Record<string, unknown>[]
  analysis_type?: SearchAnalysisType
  scope_level?: 'police_station' | 'community' | null
  scope_label?: string | null
  summary?: string | null
  row_count: number
  elapsed_ms: number
  executed_sql: string
  truncated: boolean
}

export function getReportSearchOptions() {
  return apiGet<SearchOptions>('/report-search/options')
}

export function getStatisticsDictionaryConfig() {
  return apiGet<StatisticsDictionaryConfig>('/report-search/dictionary-config')
}

export function updateStatisticsDictionaryConfig(data: {
  source: 'jjd_jjd' | 'fkd_fkd'
  disabled_categories: string[]
  disabled_types: string[]
  disabled_details: string[]
}) {
  return apiPut<StatisticsDictionarySource>('/report-search/dictionary-config', data)
}

export function getReportSearchClassifications(
  source: string,
  level: 'category' | 'type' | 'detail',
  parentCode?: string,
) {
  const params = new URLSearchParams({ source, level })
  if (parentCode) params.set('parent_code', parentCode)
  return apiGet<SearchClassificationResponse>(`/report-search/classifications?${params.toString()}`)
}

export function getReportSearchMetrics(source = 'jjd_jjd') {
  return apiGet<SearchMetrics>(`/report-search/metrics?source=${encodeURIComponent(source)}`)
}

export function executeReportSearch(data: SearchQuery) {
  return apiPost<SearchResult>('/report-search/query', data)
}

export interface BatchSearchItemResult {
  block_id: string
  success: boolean
  result?: SearchResult | null
  error?: string | null
}

export function executeReportSearchBatch(items: Array<{ block_id: string; query: SearchQuery }>) {
  return apiPost<{ items: BatchSearchItemResult[] }>('/report-search/batch-query', { items })
}

File: backend/app/schemas/report_search.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator


class SearchDepartment(BaseModel):
    code: str
    name: str


class SearchDataSource(BaseModel):
    key: str
    name: str
    enabled: bool = True


class SearchOptionResponse(BaseModel):
    current_department: SearchDepartment
    data_sources: list[SearchDataSource]
    default_start_time: datetime
    default_end_time: datetime


class SearchClassificationItem(BaseModel):
    code: str
    name: str


class SearchClassificationResponse(BaseModel):
    source: str
    level: Literal["category", "type", "detail"]
    items: list[SearchClassificationItem]


class StatisticsDictionarySource(BaseModel):
    source: str
    name: str
    categories: list[SearchClassificationItem]
    types: list[SearchClassificationItem]
    details: list[SearchClassificationItem]
    disabled: dict[Literal["category", "type", "detail"], list[str]]


class StatisticsDictionaryConfigResponse(BaseModel):
    sources: list[StatisticsDictionarySource]


class StatisticsDictionaryConfigUpdate(BaseModel):
    source: Literal["jjd_jjd", "fkd_fkd"]
    disabled_categories: list[str] = Field(default_factory=list)
    disabled_types: list[str] = Field(default_factory=list)
    disabled_details: list[str] = Field(default_factory=list)


class SearchMetricItem(BaseModel):
    key: str
    label: str
    description: str = ""
    default: bool = False


class SearchMetricResponse(BaseModel):
    source: str
    dimensions: list[SearchMetricItem]
    measures: list[SearchMetricItem]


class ReportSearchQuery(BaseModel):
    source: Literal["jjd_jjd", "fkd_fkd"] = "jjd_jjd"
    start_time: datetime
    end_time: datetime
    category_codes: list[str] = Field(default_factory=list, max_length=500)
    type_codes: list[str] = Field(default_factory=list, max_length=500)
    detail_codes: list[str] = Field(default_factory=list, max_length=500)
    dimensions: list[str] = Field(default_factory=list, max_length=3)
    measures: list[str] = Field(default_factory=lambda: ["event_count"], min_length=1, max_length=6)
    limit: int = Field(default=100, ge=1, le=500)

    @model_validator(mode="after")
    def validate_time_range(self):
        if self.end_time <= self.start_time:
            raise ValueError("结束时间必须晚于开始时间")
        if (self.end_time - self.start_time).days > 366:
            raise ValueError("单次查询时间范围不能超过 366 天")
        return self


class ReportSearchBatchItem(BaseModel):
    block_id: str = Field(..., min_length=1, max_length=100)
    query: ReportSearchQuery


class ReportSearchBatchRequest(BaseModel):
    items: list[ReportSearchBatchItem] = Field(..., min_length=1, max_length=50)


class SearchResultColumn(BaseModel):
    key: str
    label: str
    type: Literal["text", "number", "datetime"] = "text"


class ReportSearchResult(BaseModel):
    source: SearchDataSource
    department: SearchDepartment
    columns: list[SearchResultColumn]
    rows: list[dict[str, Any]]
    row_count: int
    elapsed_ms: int
    executed_sql: str
    truncated: bool = False


class ReportSearchBatchItemResult(BaseModel):
    block_id: str
    success: bool
    result: ReportSearchResult | None = None
    error: str | None = None


class ReportSearchBatchResult(BaseModel):
    items: list[ReportSearchBatchItemResult]
